### PR TITLE
MBS-7043: Add basic touch support

### PR DESCRIPTION
**Pull request by sh-csg**

### CODE REVIEW

File: amd/src/learningmap.js [hunk 1]
```diff
@@ -312,15 +317,57 @@ export const init = () => {
          * @param {*} evt
          */
         function startDrag(evt) {
-            evt.preventDefault();
+            if (evt.cancelable) {
+                evt.preventDefault();
+            }
             if (evt.target.classList.contains('learningmap-draggable')) {
+                if (evt.type == 'touchstart') {
```
Comment (PhMemmel) [inline amd/src/learningmap.js:324]: It would reduce complexity and improve readability if the touch events could be handled by separate event listener functions (could be added in the block at line 222 ff). These functions could do their touch stuff and after that call the "normal" event handler.

File: amd/src/learningmap.js [hunk 2]
```diff
@@ -515,7 +573,7 @@ export const init = () => {
     function clickHandler(event) {
         event.preventDefault();
         hideContextMenu();
-        if (event.target.classList.contains('learningmap-place') && selectedElement === null) {
+        if (event.target.classList.contains('learningmap-place') && (selectedElement === null || event.type == 'touchend')) {
```
Comment (PhMemmel) [inline amd/src/learningmap.js:576]: If going for the separate handling of touch events: Could this be realized by setting `selectedElement` to `null` in case of receiving a touchend event? Would make this a bit simpler.

**Review by PhMemmel — COMMENTED** [review]
Could only test with dev tools mobile device simulation, looks good. The time I need to wait for the modal to popup is a bit long, maybe this can be reduced a bit. I also added a suggestion to refactor the handling of touch events.

File: amd/src/learningmap.js [hunk 3]
```diff
@@ -312,15 +317,57 @@ export const init = () => {
...
+                        touchstart = true;
+                        touchmove = 0;
+                        touchend = false;
+                        setTimeout(
+                            function(evt) {
+                                if (touchmove < 3 && !touchend) {
+                                    showContextMenu(evt);
+                                }
+                            },
+                            2000,
```
[the remark this answers is not in the document]
Comment (sh-csg) [reply amd/src/learningmap.js:335]: I shortened the timeout - but this might interfere with dragging the places. Maybe we have to wait for user feedback on this.

File: amd/src/learningmap.js (hunk 1 above)
Comment (sh-csg) [reply amd/src/learningmap.js:324]: Done

File: amd/src/learningmap.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sh-csg) [reply amd/src/learningmap.js:335]: Unfortunately a shorter timeout interferes a lot with dragging and double clicking.

File: amd/src/learningmap.js (hunk 2 above)
Comment (sh-csg) [reply amd/src/learningmap.js:595]: Done

File: amd/src/learningmap.js [hunk 4]
```diff
@@ -373,11 +383,78 @@ export const init = () => {
...
+                    touchmove = 0;
+                    touchend = false;
+                    setTimeout(
+                        (evt) => {
+                            if (touchmove < 3 && !touchend) {
+                                showContextMenu(evt);
+                            }
+                        },
+                        2000,
+                        evt
```
Comment (PhMemmel) [inline amd/src/learningmap.js:414]: is this additional param needed?

File: amd/src/learningmap.js [hunk 5]
```diff
@@ -373,11 +383,78 @@ export const init = () => {
...
+            } else {
+                if (!touchstart) {
+                    touchstart = true;
+                    touchend = false;
+                    touchmove = 0;
+                    setTimeout(
+                        () => {
+                            touchstart = false;
+                        },
+                    300 );
```
Comment (PhMemmel) [inline amd/src/learningmap.js:435]: remove space?

File: amd/src/learningmap.js [hunk 6]
```diff
@@ -373,11 +383,78 @@ export const init = () => {
...
+                            }
+                        },
+                        2000,
+                        evt
+                    );
+                    setTimeout(
+                        () => {
+                            touchstart = false;
+                        },
+                    300 );
```
Comment (PhMemmel) [inline amd/src/learningmap.js:420]: remove space?

File: amd/src/learningmap.js [hunk 7]
```diff
@@ -373,11 +383,78 @@ export const init = () => {
...
+                }
+            }
+        }
+
+        /**
+         * function called when touchend, touchleave or touchcancel event occurs.
+         * @param {*} evt
+         */
+        function endTouch(evt) {
+            if (touchmove < 3 && touchstart) {
```
Comment (PhMemmel) [inline amd/src/learningmap.js:448]: I think a comment explaining what `touchmove < 3` is for would be nice

File: amd/src/learningmap.js (hunk 2 above)
Comment (PhMemmel) [reply amd/src/learningmap.js:595]: That's still present?

File: amd/src/learningmap.js [hunk 8]
```diff
@@ -22,6 +22,11 @@ export const init = () => {
     // Variable for storing the selected element for the activity selector
     var elementForActivitySelector = null;
 
+    // Variables for simulating double click on touch devices
+    var touchstart = false;
+    var touchmove = 0;
```
Comment (PhMemmel) [inline amd/src/learningmap.js:27]: A comment would be nice explaining what this is for

File: amd/src/learningmap.js [hunk 9]
```diff
@@ -330,7 +337,10 @@ export const init = () => {
          * @param {*} evt
          */
         function drag(evt) {
-            evt.preventDefault();
+            if (evt.cancelable) {
+                evt.preventDefault();
+            }
+            touchmove++;
```
Comment (PhMemmel) [inline amd/src/learningmap.js:343]: A comment why this is being incremented would be nice

File: amd/src/learningmap.js (hunk 4 above)
Comment (sh-csg) [reply amd/src/learningmap.js:414]: To hand over the event to the called function as a parameter

**Comment by sh-csg** [comment]
All requested changes are done, branch is rebased on latest master. Moving version number to 0.8.

**Comment by sh-csg** [comment]
Also changed supported moodle versions (due to 4.1 and as a separate branch for 3.9/3.10 needs to be created).

File: amd/src/learningmap.js (hunk 4 above)
Comment (PhMemmel) [reply amd/src/learningmap.js:418]: Yes, I know what the param means, I just wanted to ask, if it's needed there. But I tested (mainly out of personal interest ;-)) and you're totally right. It has to be passed there to enable the arrow function to pass it onto `showContextMenu`. Sorry for asking :-)

File: amd/src/learningmap.js [hunk 10]
```diff
@@ -285,11 +292,11 @@ export const init = () => {
...
             el.addEventListener('mouseleave', endDrag);
-            el.addEventListener('touchstart', startDrag);
+            el.addEventListener('touchstart', startTouch);
             el.addEventListener('touchmove', drag);
-            el.addEventListener('touchend', endDrag);
-            el.addEventListener('touchleave', endDrag);
-            el.addEventListener('touchcancel', endDrag);
+            el.addEventListener('touchend', endTouch);
+            el.addEventListener('touchleave', endTouch);
+            el.addEventListener('touchcancel', endTouch);
```
Comment (PhMemmel) [inline amd/src/learningmap.js:299]: This is much more cleaner now, gj!

**Comment by PhMemmel** [comment]
From code perspective it now looks good. I also tested again via browser dev tools, could not find any issues. This should be tested as soon as possible on a real mobile device. Just a minor thought:

The context menu is always being shown in the left upper corner now. Is this a mobile device optimization or is it intended to open it at the point where I "click" on a touch device?

And another thought which comes to my mind right now: Maybe one could circumvent the usage of the boolean variables to detect "fake touch double click" etc. by using promises. But that's just a quick thought, I do not have a precise prospect on how this would work. Thoughts? However, current code is clean, understandable and stable, so this definitely does not need to be implemented for this.

Good job!

**Comment by sh-csg** [comment]
Changing the way of implementation may be a task for a later version - I can't handle that in the current PR. Positioning of the modal is fixed now.

**Comment by PhMemmel** [comment]
Yes, as I wrote: Not intended for this PR. Will double check positioning later.

**Comment by PhMemmel** [comment]
Unfortunately, the last commit makes the context menu appear outside the screen on some small devices or out of the editor area which needs to be fixed. If this is too complicated, I think we also can live with the version before where it always appears in the left upper corner.

**Comment by PhMemmel** [comment]
I think what we want for the context menu would be a behavior like in this demo. Maybe the solution provided there could also be implemented: https://mionskowski.pl/posts/positioning-a-context-menu-using-pure-css/

**Comment by sh-csg** [comment]
Thanks a lot for the suggestion. I changed the way of implementation in the last commit:
The context menu appearing also outside of the window remains as an issue (but one has to decide whether this is really an issue). But it stays always inside the viewport.
position: fixed leads to another (new) issue when scrolling as the context menu keeps position.

**Comment by PhMemmel** [comment]
Looks good in general. However, on small devices when opening the context menu for a place which is pretty much at the right edge of the map there is an overflow. Did you mean that with 'outside of the window'? Can this be easily fixed?

image

**Comment by sh-csg** [comment]
I was not able to reproduce that - but maybe this small change helps

**Comment by PhMemmel** [comment]
Tested again based on the last commit: It seems to solve all the issues with the positioning of the modal, good job!

Also tested basic functionality: Adding and removing places, moving places around, opening modal (positioning now is great, also in all edge cases I tested). Could not test on *real* mobile devices, but in Chrome Dev Tools everything looks good.
